### DISCUSSION TITLE
Can I create an array union type from a union type without arrays?

### DISCUSSION THREAD

**Discussion by anthonyma94** [colinhacks/zod#1601] [Q&A] [answered]
The Zod alternative I'm looking for is this, which creates a union type where each type is converted into an array:
```ts
type UnionArraysFromUnionTypes<T> = T extends infer I ? I[] : never;
type PrimitiveValues = string | number;

PrimitiveValues[]; // (string | number)[];
UnionArraysFromUnionTypes<PrimitiveValues>; // string[] | number[]
```

Is there a Zod way of doing this?

**Comment by JacobWeisenburger** [comment]
There might be a few different ways to do this, but it would help someone to answer your question if you provided a bit more of your use case.

**Reply by anthonyma94** [reply]
I have a union type `z.union([z.string(), z.number()])`, I'd like to create type `z.union([z.string().array(), z.number().array()])`, where the second array will always be a union type made up of each of the first union's type as an array. It has to be a discriminating union, the array must only have a single type in it. I'd like to do this by referencing the first type somehow, so if I need to add another type in the union I don't have to do it in 2 places.

**Comment by JacobWeisenburger** [comment, marked as answer]
I think the main problem is that `z.union` expects a tuple with at least 2 elements. So when using `array.map`, you need to use `as` to tell typescript that the new array will have at least 2 elements.

```ts
const schema = z.union( [ z.string(), z.number() ] )
type Schema = z.infer<typeof schema>
// type Schema = string | number

console.log( schema.parse( 'foo' ) ) // foo
console.log( schema.parse( 42 ) ) // 42

const newSchema = z.union(
    schema.options.map( x => x.array() ) as [
        z.ZodArray<z.ZodString>,
        z.ZodArray<z.ZodNumber>,
    ]
)
type NewSchema = z.infer<typeof newSchema>
// type NewSchema = string[] | number[]

console.log( newSchema.parse( [ '1', '2', '3' ] ) ) // [ '1', '2', '3' ]
console.log( newSchema.parse( [ 1, 2, 3 ] ) ) // [ 1, 2, 3 ]
```
